### PR TITLE
Event API - Agreement events

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**
resolves: #729

AgreementTerminatedEvent timestamp is return as String, because of bug in ya-client.
Needs fix: https://github.com/golemfactory/ya-client/pull/64

### CODE REVIEW

File: core/market/src/rest_api/common.rs [hunk 1]
```diff
@@ -41,3 +45,28 @@ async fn get_agreement(
...
+    query: Query<QueryAgreementEvents>,
+    id: Identity,
+) -> impl Responder {
+    let timeout: f32 = query.timeout;
+    let after_timestamp = query
+        .after_timestamp
+        .unwrap_or(Utc.ymd(1970, 1, 1).and_hms(0, 0, 0));
+
+    market
+        .requestor_engine
```
Comment (tworec) [inline core/market/src/rest_api/common.rs:61]: move to `market`

File: core/market/src/rest_api/common.rs [hunk 2]
```diff
@@ -41,3 +45,28 @@ async fn get_agreement(
...
+#[actix_web::get("/agreements/events")]
+async fn collect_agreement_events(
+    market: Data<Arc<MarketService>>,
+    query: Query<QueryAgreementEvents>,
+    id: Identity,
+) -> impl Responder {
+    let timeout: f32 = query.timeout;
+    let after_timestamp = query
+        .after_timestamp
+        .unwrap_or(Utc.ymd(1970, 1, 1).and_hms(0, 0, 0));
```
Comment (tworec) [inline core/market/src/rest_api/common.rs:58]: use golem crowfunding date as `zero` time

File: core/market/src/negotiation/provider.rs [hunk 3]
```diff
@@ -194,18 +197,19 @@ impl ProviderBroker {
...
             AgreementState::Proposal => AgreementStateError::Proposed(agreement.id),
             AgreementState::Pending => {
-                // TODO : possible race condition here ISSUE#430
+                // TODO: possible race condition here ISSUE#430
                 // 1. this state check should be also `db.update_state`
                 // 2. `db.update_state` must be invoked after successful propose_agreement
+                // TODO: if dao.approve fails, Provider and Requestor have inconsistent state.
                 self.api.approve_agreement(agreement, timeout).await?;
-                dao.update_state(agreement_id, AgreementState::Approved)
+                dao.approve(agreement_id)
```
Comment (tworec) [inline core/market/src/negotiation/provider.rs:205]: we need to notify agreement events here

File: core/market/src/negotiation/requestor.rs [hunk 4]
```diff
@@ -34,26 +39,22 @@ pub enum ApprovalStatus {
...
     api: NegotiationApi,
-    agreement_notifier: EventNotifier<AgreementId>,
 }
 
 impl RequestorBroker {
     pub fn new(
         db: DbExecutor,
         store: SubscriptionStore,
         proposal_receiver: UnboundedReceiver<RawProposal>,
+        session_notifier: EventNotifier<AppSessionId>,
```
Comment (tworec) [inline core/market/src/negotiation/requestor.rs:49]: why do we need it to be common?

**Review by tworec — COMMENTED** [review]
(initial) live review comments

File: core/market/Cargo.toml [hunk 5]
```diff
@@ -12,6 +12,7 @@ testing = ["actix-http", "actix-rt", "actix-service", "env_logger"]
 
 [dependencies]
 ya-agreement-utils = "0.1"
+ya-diesel-utils = { version = "0.1", path = "../../utils/diesel-utils" }
```
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/Cargo.toml:15]: Fixed in https://github.com/golemfactory/yagna/pull/793

File: core/market/src/rest_api/common.rs (hunk 1 above)
Comment (nieznanysprawiciel) [reply core/market/src/rest_api/common.rs:61]: Fixed in #793

File: core/market/src/rest_api/common.rs (hunk 2 above)
Comment (nieznanysprawiciel) [reply core/market/src/rest_api/common.rs:58]: Fixed in #793

File: core/market/src/negotiation/provider.rs (hunk 3 above)
Comment (nieznanysprawiciel) [reply core/market/src/negotiation/provider.rs:205]: Fixed in #793

**Comment by nieznanysprawiciel** [comment]
Merged as:  #793

**Pull request closed without merging**